[RFR] Create DeliveryData object, refactor DeliveryData dependencies

SimonBiggs opened the pull request:
Preparing to expose `trf2dcm` and the mudensity script.

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/exportable.py:
```diff
@@ -0,0 +1,124 @@
+
+
+class DeliveryData(DeliveryDataBase):
+    def __new__(cls, *args):
```
@Centrus007 here is the new DeliveryData class that is worth preemptively reviewing.

SimonBiggs commented:
@Centrus007 actually, I've changed my mind. It's be great to have this one merged in as it is. I'm going to continue work in another branch.

Let me go through and make some dot points.

SimonBiggs commented on packages/pymedphys/src/pymedphys/deliverydata.py:
```diff
@@ -1 +1,4 @@
-from pymedphys_core.deliverydata import DeliveryData
+from pymedphys_deliverydata import DeliveryData
+from pymedphys_deliverydata.dicom import (
```
I deleted pymedphys_core and instead made pymedphys_mudensity and pymedphys_deliverydata

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/base/core.py:
```diff
@@ -0,0 +1,35 @@
+# Make the creation of DeliveryData look a bit like this...
+# Or a bit like the dataclass ...
+DeliveryDataBase = namedtuple(
+    'DeliveryData',
```
The DeliveryData object is still a namedtuple at its core.

DeliveryDataBase is used throughout the pymedphys_deliverydata so as to avoid an internal circular dependency. Then, at a top level of pymedphys_deliverydata the functions within the module are attached to the primary DeliveryData object which is then used by every other package.

SimonBiggs replied:
As you can see, DeliveryDataBase is only used internally:

![image](https://user-images.githubusercontent.com/6559099/58230214-0893e080-7d77-11e9-9416-b59647c22ec2.png)

...in fact, maybe I should prepend it with _

Centrus007 replied:
I'm not sure I see the circular dependency issue here. Can you explain what code specifically depends on this, and why it would in turn depend on that code if not separated as a "Base" object? Is it an artefact of preferring functions over methods? If so... why not stick to methods? :p

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/exportable.py:
```diff
@@ -0,0 +1,124 @@
+        return filter_out_irrelevant_control_points(self)
+
+    @functools.lru_cache()
+    def mask_by_gantry(self, angles: Union[tuple, float, int], tolerance=3):
```
Function results are cached, so repeat calls just look up the cache.

Centrus007 replied:
Cool (I think)! Never seen this.

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/exportable.py:
```diff
@@ -0,0 +1,124 @@
+
+        # TODO: This is worth discussion whether or not we should do this
+        if len(mudensities) == 1:
+            return mudensities[0]
```
If one or no gantry angle is provided mudensity is returned as a numpy array.

If more than one gantry angle is provided mudensity is returned as a list of numpy arrays.

Do you think this is a good idea?

Centrus007 replied:
I think I'd slightly prefer the type to be consistent, even if it does mean that the user has to go through the tedium of `mudensities[0]`

Also, imagine some sort of batch processing that loops through the returned `mudensities`. You don't want code intended to loop through a list of mudensities (that theoretically could support a list with a single item) to instead loop through the numpy array...

SimonBiggs replied:
Yup, that's what I was thinking. As I wrote it I went ... hmmm.... not sure.

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/utilities/__init__.py:
```diff
@@ -24,16 +24,11 @@
-
-Example:
-    >>> from pymedphys.deliverydata import DeliveryData
-"""
```
The docs are going to be a little bit in disarray...

SimonBiggs commented on packages/pymedphys_deliverydata/tests/deliverydata/test_dicom_round_trip.py:
```diff
@@ -26,34 +26,81 @@
+
+@pytest.fixture
+def logfile_delivery_data() -> DeliveryData:
+    return delivery_data_from_logfile(LOGFILE_FILEPATH)
```
Just putting it out there, fixtures are awesome! ... I am hooked...

Centrus007 replied:
Agree!

SimonBiggs commented on packages/pymedphys_deliverydata/tests/deliverydata/test_dicom_round_trip.py:
```diff
@@ -26,34 +26,81 @@
+        raise
+
+
+def test_mudensity_agreement(loaded_dicom_dataset, logfile_delivery_data):
```
Here are some tests for the trf2dcm code

SimonBiggs commented on packages/pymedphys_deliverydata/tests/exportable/test_exported_object.py:
```diff
@@ -0,0 +1,10 @@
+    empty = DeliveryData.empty()
+    filtered = empty.filter_cps()
+
+    assert type(filtered.monitor_units) is tuple
```
Here is a little test I wrote during trouble shooting

SimonBiggs commented on packages/pymedphys_dicom/src/pymedphys_dicom/rtplan/build.py:
```diff
@@ -0,0 +1,89 @@
+
+        cps.append(add_data_to_control_point(template, data, i))
+
+    return cps
```
The beginnings of a generic RT Plan generation library :)

Centrus007 replied:
This is very, very cool!

SimonBiggs replied:
:)

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/dicom/conversion.py:
```diff
@@ -29,62 +29,49 @@
+    build_control_points,
+    replace_fraction_group,
+    replace_beam_sequence,
+    restore_trailing_zeros)
```
I've decided to follow suite with your convention of keeping packages each on a new line. That means changing a package only is clearer with version control.

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/__init__.py:
```diff
@@ -0,0 +1 @@
+from .exportable import DeliveryData
```
DeliveryData is importable at a top level:

```python
from pymedphys_deliverydata import DeliveryData 
```

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/exportable.py:
```diff
@@ -0,0 +1,124 @@
+from ..utilities import (
+    filter_out_irrelevant_control_points,
+    get_all_masked_delivery_data,
+    get_metersets_from_delivery_data)
```
Notice here how I am only ever importing from the module directories. I am trying to keep a separation of concerns here. Each directory module decides what API it will expose to the big bad world, that way, other modules in the same package are restricted to only using what that module is exposing.

Keeps a nice well defined API boundary.

Centrus007 replied:
I see. I was wondering why these files were separated into subfolders. Now I know :)

SimonBiggs commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/exportable.py:
```diff
@@ -0,0 +1,124 @@
+            self.mask_by_gantry(gantry_angles, gantry_tolerance))
+
+    def mudensity(self, gantry_angles=None, gantry_tolerance=None, grid_resolution=1):
+        if gantry_angles is None:
```
This method... this method is a pleasure to use... :)

SimonBiggs commented:
Passes :) (yayy)

SimonBiggs commented:
Ahh, don't merge, the docs don't look like they're building.

SimonBiggs commented:
@Centrus007 alright! Docs are building now.

Two new issues have been made to track what happened there:

https://github.com/pymedphys/pymedphys/issues/341

https://github.com/pymedphys/pymedphys/issues/340

SimonBiggs commented:
@Centrus007 when you do get the chance, this is now ready for review and merge :).

Centrus007 commented:
To be honest mate, this is still very convoluted (and I'm struggling to see why, to be honest). It could really do with docstrings, and maybe even your seemingly-most-hated thing: comments!

Centrus007 commented:
Why are `conversion` and `utilities` separate files?

Centrus007 commented:
How does each of the plethora of functions (dare I say too many that are too small - and I strongly believe there IS such a thing) related to one another.

Functionalising everything so heavily obfuscates logical connections between code (since it becomes far more physically separated than logically separated).

SimonBiggs commented:
> Why are `conversion` and `core` and 'masking` separate files?

More than happy for you to recommend an adjustment

SimonBiggs commented:
> How does each of the plethora of functions (dare I say too many that are too small - and I strongly believe there IS such a thing) relate to one another.

You're going to need to give me some examples

SimonBiggs commented:
Certainly a few of the functions within `conversion` can be moved over to the `utilities` module.

SimonBiggs commented:
> To be honest mate, this is still very convoluted (and I'm struggling to see why, to be honest). It could really do with docstrings, and maybe even your seemingly-most-hated thing: comments!

Fair points.

SimonBiggs commented:
> Functionalising everything so heavily obfuscates logical connections between code (since it becomes far more physically separated than logically separated).

Maybe this can be spun off into an issue separate from this pull request? What do you think?

Centrus007 commented:
> Functionalising everything so heavily obfuscates logical connections between code (since it becomes far more physically separated than logically separated).

This was probably a premature statement, to be honest. I'm still more use to reviewing OOP code, so I know where to look for top-level functions and progressively delve into deeper levels from there. But I eventually found my way around.

There might be something to be said for thinking about how we sort functions within a module.

But mainly, I think docstrings help.

I shouldn't really even be reviewing this - my frustration from a late night at work troubleshooting machine issues is probably leaking out. I'm sorry if it has, and I think I'd better call it a night soon :)

SimonBiggs commented:
> There might be something to be said for thinking about how we sort functions within a module.

I completely agree. That would certainly help.

Centrus007 commented:
stacktraces too, now that I think about it...

SimonBiggs commented:
> stacktraces too, now that I think about it...

?

SimonBiggs commented:
> my frustration from a late night at work troubleshooting machine issues is probably leaking out. I'm sorry if it has, and I think I'd better call it a night soon :)

All good :). I must say I had the wind taken out of my sails a little

Centrus007 commented on app/src/python/user-scripts/trf2dcm.py:
```diff
@@ -0,0 +1,32 @@
+output_directory = 'output'
+
+trf_filepaths = glob(os.path.join(input_directory, '*.trf'))
+first_dicom_filepath = glob(os.path.join(input_directory, '*.dcm'))[0]
```
This is an interesting way to do it. The user provides a DICOM as a template, basically?

Is the user given instructions to that effect? Haven't found them so far, but I might have just missed them!

SimonBiggs replied:
The user script isn't live yet :). I have commented out the option. Left it for another pull request.

SimonBiggs replied:
But yes, that's right. I've explained things on the app page:

![image](https://user-images.githubusercontent.com/6559099/58260169-52051f80-7db9-11e9-952a-46e117989adf.png)

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/dicom/conversion.py:
```diff
@@ -29,62 +29,49 @@
-    closest_left_leaning = not_180[where_closest_left_leaning]
-    closest_right_leaning = not_180[where_closest_right_leaning]
+def gantry_tol_from_gantry_angles(gantry_angles):
+    min_diff = np.min(np.diff(sorted(gantry_angles)))
```
What is this gantry tolerance (used for)? Can you document it?

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/dicom/conversion.py:
```diff
@@ -29,62 +29,49 @@
-    angle[is_180] = np.sign(angle[closest_left_leaning]) * angle[is_180]
 
-    return angle
+def delivery_data_to_dicom(delivery_data: _DeliveryDataBase, dicom_template):
```
Not `from_dicom()` and `to_dicom()` methods instead?

SimonBiggs replied:
This does become from_dicom and to_dicom in the top level class.

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/dicom/conversion.py:
```diff
@@ -97,7 +84,7 @@ def dicom_to_delivery_data(dicom_dataset) -> DeliveryData:
 
 
-def merge_delivery_data(separate: List[DeliveryData]) -> DeliveryData:
+def merge_delivery_data(separate: List[_DeliveryDataBase]) -> _DeliveryDataBase:
```
Could do with an explanation of this one one day!

SimonBiggs replied:
The type hints?

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/exportable.py:
```diff
@@ -0,0 +1,130 @@
+            dicom_to_delivery_data(dataset))
+
+    @classmethod
+    def empty(cls):
```
What's the use case?

If it's effectively a whole bunch of (immutable) tuples of and tuples of tuples... would you ever make an empty one?

SimonBiggs replied:
I used it for testing...

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/exportable.py:
```diff
@@ -0,0 +1,130 @@
+            ))
+        )
+
+    def to_dicom(self, template):
```
Ah, you do have a `to_dicom()`.

Is it really necessary to maintain the function? (This is kinda what I mean by functionalised code, I now need to look elsewhere to find out what's going on here - in a different FILE even :/)

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/utilities/core.py:
```diff
@@ -66,35 +59,25 @@ def extract_angle_from_delivery_data(delivery_data, gantry_angle,
-    """Removes control points that don't have MU delivery
-    """
-    assert len(mu) > 0, "No control points found"
+def filter_out_irrelevant_control_points(delivery_data: _DeliveryDataBase) -> _DeliveryDataBase:
```
This was **BORN** to be a *method* of `DeliveryData`! It causes a simple state change of a `DeliveryData` object. This is very ugly (to me) as a function...

SimonBiggs replied:
Thing is, it can run on anything that looks like a delivery data object... if it was a method only, that wouldn't be possible.

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/utilities/masking.py:
```diff
@@ -0,0 +1,124 @@
+    return near_angle
+
+
+def apply_mask_to_delivery_data(delivery_data: _DeliveryDataBase, mask):
```
method

SimonBiggs replied:
Also... this function was written before the class was decided to be made.

... if that helps see my reasoning.

Centrus007 commented on packages/pymedphys_dicom/src/pymedphys_dicom/rtplan/core.py:
```diff
@@ -0,0 +1,68 @@
+from pymedphys_utilities.transforms import convert_angle_to_bipolar
+
+
+def get_metersets_from_dicom(dicom_dataset, fraction_group):
```
Not a strong preference, but the word "get" in these functions is a little superfluous and arguably unpythonic. Admittedly it's getter and setter *methods* that are generally unpreferred, but this function was also **BORN** to be a method of a dicom object :D

SimonBiggs replied:
Fair point. I believe I agree.

Centrus007 commented on packages/pymedphys_logfiles/src/pymedphys_logfiles/elekta/bygantry.py:
```diff
@@ -148,13 +141,17 @@ def calc_logfile_mu_density_bygantry(index, config, logfile_group,
 
 
 def compare_logfile_group_bygantry(index, config, cursor, logfile_group,
```
```suggestion
def compare_logfile_group_by_gantry_angle(index, config, cursor, logfile_group,
```

SimonBiggs replied:
This change would need to be propagated... so can't just accept this here

Centrus007 commented on packages/pymedphys_mudensity/src/pymedphys_mudensity/mudensity/core.py:
```diff
@@ -29,9 +29,9 @@
-from ..deliverydata import (
-    extract_angle_from_delivery_data, remove_irrelevant_control_points)
+from pymedphys_utilities.rtplan import (
+    find_relevant_control_points,
```
Also a born-to-be-a-method function :p

SimonBiggs replied:
:) Class needs to be made first :)

Centrus007 commented on packages/pymedphys_utilities/src/pymedphys_utilities/transforms/bipolar.py:
```diff
@@ -0,0 +1,56 @@
+import numpy as np
+
+
+def convert_angle_to_bipolar(angle):
```
Best to have a more descriptive function name. What coordinate system is the original angle in?

Centrus007 replied:
(Probably IEC gantry...)

Centrus007 commented:
> All good :). I must say I had the wind taken out of my sails a little, but although I care about my work, I certainly do, I have certainly learnt to take much criticism :)

I hear you.

I think I have a strong negative bias when reviewing - I'm trying to look for areas of improvement. I should remember to give praise where it's due (and with you, it's always, always, ALWAYS due!) I will work on that.

This is a fantastic piece of work - I'm really looking forward to properly diving in and wish I had earlier. It's so much better than other implementations I've seen.

SimonBiggs commented:
@Centrus007 would you be open to us spinning out these review comments into a single issue with the aim to address them after the conference? Then merge this in as is?

SimonBiggs commented:
Another justification for this... Is if you compare DeliveryData before (and all the functions that call it) to what it is now...

SimonBiggs commented:
By the way, most of your comments here aren't actually on this pull request. This pull request was the first iteration of the DeliveryData class + updating and moving functions that are already in the code base to support and use that class...

Centrus007 commented:
> @Centrus007 would you be open to us spinning out these review comments into a single issue with the aim to address them after the conference? Then merge this in as is?

Yes, sorry. Is there an easy, built-in way to spin out comments?

> By the way, most of your comments here aren't actually on this pull request. This pull request was the first iteration of the DeliveryData class + updating and moving functions that are already in the code base to support and use that class...

> As in, most of the functions here were already in the code base... (And need work, I agree, just not today, not in this pull request...)

I can wait :) Besides, it might be a good exercise in familiarisation for me to have a go after the conference (with a view to including Varian logfile support)

Now I think about it, I actually think this code is old enough to have never been reviewed...

Centrus007 commented:
Given that we are under time pressure for the code in this PR, I think I'm happy. But I don't want to make a habit of premature merges. PRs are our only explicitly triggered opportunities to review code. If code makes it through with unperformed changes intended for the future, what triggers those future changes? Probably nothing until we implement a new feature...

It's okay to iterate within a PR instead of within `master`.

SimonBiggs commented:
> It's okay to iterate within a PR instead of within `master`.

Fair point :)

> Spinning out comments into an issue at least maintains a semi-visible record, but this record is buried in over 100 other issues with no trigger to address the comments. They're inevitably deprioritised in favour of other things.

Yup, certainly.

I'm implementing a range of your suggestions as I bump into them while I am implementing trf2dcm user script over at https://github.com/pymedphys/pymedphys/pull/346

I am also making it so that merge request is pulling into this one.

Centrus007 commented:
> I am also making it so that merge request is pulling into this one, that way, the pull request diff is compared to this, not to master. So you only have to read the new changes.

Did you still want to merge this one in then? Or wait?

SimonBiggs commented:
> > I am also making it so that merge request is pulling into this one, that way, the pull request diff is compared to this, not to master. So you only have to read the new changes.
> 
> Did you still want to merge this one in then? Or wait?

Based on the following:

> Spinning out comments into an issue at least maintains a semi-visible record, but this record is buried in over 100 other issues with no trigger to address the comments. They're inevitably deprioritised in favour of other things.

I'd like to wait at least a bit longer, let's put it through another pass once I have finished this part I'm working on.

SimonBiggs commented:
Just merged that other one in, because I realised the following link can show you the new changes :)

http://github.com/pymedphys/pymedphys/compare/8658e65...e77a526

SimonBiggs commented:
@Centrus007 I certainly didn't hit everything you were after, no where near close. But I did implement a few suggestions.

So this is ready for review with the following changes since last time:

https://github.com/pymedphys/pymedphys/compare/8658e65...e77a526

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/dicom/to_dicom.py:
```diff
@@ -0,0 +1,160 @@
+from ..utilities import get_all_masked_delivery_data
+
+
+def delivery_data_to_dicom(filtered_delivery_data: DeliveryDataBase,
```
Why does this need to be a separate function? I really think it's overgeneralising at the expense of readability. I think it is vanishingly unlikely that this function will (or at least *should*) ever be used on a `DeliveryDataBase`-like object...

SimonBiggs replied:
Fair point... can we leave it like this for a bit? And let it settle. I think it will be worth revisiting delivery data after the DICOM objects are built...

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/dicom/to_dicom.py:
```diff
@@ -0,0 +1,160 @@
+    }
+
+
+def jaw_dd2dcm(jaw):
```
`dd2dcm`? This is sooooo un-SimonBiggsy.

And you complain about my preference for `ds`???

(you don't actually complain, I'm playing)

SimonBiggs replied:
:)

Centrus007 commented on packages/pymedphys_deliverydata/src/pymedphys_deliverydata/logfile/trf.py:
```diff
@@ -41,38 +42,37 @@ def delivery_data_from_logfile(logfile_path):
-def delivery_data_from_pandas(logfile_dataframe):
-    raw_monitor_units = logfile_dataframe[
-        'Step Dose/Actual Value (Mu)'].values.tolist()
+def delivery_data_from_pandas(logfile_dataframe) -> DeliveryDataBase:
```
See point made above about need for separate function...

SimonBiggs replied:
In this case I disagree more strongly. This pandas dataframe format is very, very specific to the trf logfile format.

SimonBiggs commented on app/src/python/user-scripts/trf2dcm.py:
```diff
@@ -0,0 +1,31 @@
+    delivery_data = DeliveryData.from_logfile(filepath)
+
+    print("Converting log file to RT Plan DICOM")
+    created_dicom = delivery_data.to_dicom(dicom_template)
```
I'll hold off this suggestion here. Once we have a nice DICOM object I think we'll align to something clearer throughout the codebase.

SimonBiggs commented on app/src/python/user-scripts/trf2dcm.py:
```diff
@@ -0,0 +1,31 @@
+trf_filepaths = glob(os.path.join(input_directory, '*.trf'))
+first_dicom_filepath = glob(os.path.join(input_directory, '*.dcm'))[0]
+
+dicom_template = pydicom.dcmread(first_dicom_filepath, force=True)
```
:)

SimonBiggs replied:
It's also a template DICOM header...

SimonBiggs commented:
Once this clears CI I think I'll be ready to merge...

SimonBiggs commented:
Gee... the npm package server is very flakey at the moment... gee I am glad pypi doesn't have these issues.
